### DISCUSSION TITLE
AIX library load issue

### DISCUSSION THREAD

**Discussion by hurzhurz** [pyinstaller/pyinstaller#7933] [Help]
Hi,
I'm triying to build saltstack as onedir-bundle using [tiamat](https://gitlab.com/saltstack/pop/tiamat) and PyInstaller under AIX.
For the most part this works very well, but I have some issues with libraries and could use some help.

When I run the main binary on the same machine where I built it, it works with no problem.

But on a second AIX machine, where the system libraries slightly differ and where python is not installed, I ran into an issue:

```
root@aixtest2:/opt/salt-test>run-default/run call
[33030558] Error loading Python lib '/opt/salt-test/run-default/libpython3.9.a(libpython3.9.so)': dlopen:       0509-130 Symbol resolution failed for /opt/salt-test/run-default/libpython3.9.a[libpython3.9.so] because:
        0509-136   Symbol _GLOBAL__AIXI_libintl_so (number 284) is not exported from
                   dependent module /opt/freeware/lib/libintl.a[libintl.so.8].
        0509-136   Symbol _GLOBAL__AIXD_libintl_so (number 285) is not exported from
                   dependent module /opt/freeware/lib/libintl.a[libintl.so.8].
        0509-022 Cannot load module /opt/salt-test/run-default/libpython3.9.a(libpython3.9.so).
        0509-026 System error: Cannot run a file that does not have a valid format.
        0509-192 Examine .loader section symbols with the
                 'dump -Tv' command.
```
So the bundled `libpython3.9.a` doesn't like the systems `libintl.a`.
As I found out, libintl.a is not included in the bundle, probably because of an exclude list in [dylib.py](https://github.com/pyinstaller/pyinstaller/blob/develop/PyInstaller/depend/dylib.py#L225).

For testing I modified dylib.py and added libintl.a to `_includes`.
This helped a bit. libintl.a is included in the bundle and loaded.
But now I get this:
```
root@aixtest2:/opt/salt-test>run-withlibintl/run call
[31457734] Error loading Python lib '/opt/salt-test/run-withlibintl/libpython3.9.a(libpython3.9.so)': dlopen:   0509-130 Symbol resolution failed for /opt/salt-test/run-withlibintl/libintl.a[libintl.so.8] because:
        0509-136   Symbol _GLOBAL__AIXI_libiconv_so (number 4) is not exported from
                   dependent module /opt/freeware/lib/libiconv.a[libiconv.so.2].
        0509-136   Symbol _GLOBAL__AIXD_libiconv_so (number 5) is not exported from
                   dependent module /opt/freeware/lib/libiconv.a[libiconv.so.2].
        0509-022 Cannot load module /opt/salt-test/run-withlibintl/libpython3.9.a(libpython3.9.so).
        0509-026 System error: Cannot run a file that does not have a valid format.
        0509-192 Examine .loader section symbols with the
                 'dump -Tv' command.
```
So `libintl.a` seems to need `libiconv.a` and tries to load it from the systems library path /opt/freeware/lib/.
But libiconv.a is and was also included in the bundle.
And if I temporarily replace the one in /opt/freeware/lib/ with the one from the bundle, everything works. But of course, this is not a solution.

How can I get it to load libiconv.a from the bundle directory?

Any hint on what I could try would be great!

In case it helps, here are the build logs and and the spec files (generated by tiamat), with dylib.py modification (to include libintl) and without:
[default.log](https://github.com/pyinstaller/pyinstaller/files/12597973/default.log)
[salt.spec.default](https://github.com/pyinstaller/pyinstaller/files/12597974/salt.spec.default.txt)
[with-libintl.log](https://github.com/pyinstaller/pyinstaller/files/12597976/with-libintl.log)
[salt.spec.withlibintl](https://github.com/pyinstaller/pyinstaller/files/12597975/salt.spec.withlibintl.txt)

**Comment by rokm** [comment]
Hmmm, can't help you much with AIX, I'm afraid.

But just to clarify, is the second AIX machine running the same version of the OS as the first one? If the second one is older, I would not be surprised if there are missing symbols in the libraries collected from the newer systems (same happens on linux with glibc, or on macOS with system libraries). Or even if the second machine has newer version of AIX, and `libintl` is of different version with incompatible ABI.

Does the collected `libintl` have some sort of run-path (or equivalent set)? That might force it to prefer the `libiconv` in `/opt/freeware/lib` instead of the one collected in the application bundle...

**Reply by hurzhurz** [reply]
> But just to clarify, is the second AIX machine running the same version of the OS as the first one? If the second one is older, I would not be surprised if there are missing symbols in the libraries collected from the newer systems (same happens on linux with glibc, or on macOS with system libraries). Or even if the second machine has newer version of AIX, and `libintl` is of different version with incompatible ABI.

I'm actually not that familiar with AIX either. But as far as I can tell, both have the same OS version (oslevel 7.2.0.0).
But on the build-machine, I installed the [AIX Toolbox for Open Source Software](https://www.ibm.com/support/pages/node/882892) so I can use DNF to install gcc, python, etc. for building.
And this brings different/newer versions of some libraries with it. I don't know whats the source of the libraries on the second machine.

> Does the collected `libintl` have some sort of run-path (or equivalent set)? That might force it to prefer the `libiconv` in `/opt/freeware/lib` instead of the one collected in the application bundle...

Hmm... I don't know much about the internals of library files. But maybe you mean something like this?:
```
> strings libintl.a  | grep freeware
/opt/freeware/share/locale
/opt/freeware/lib64:/opt/freeware/lib:/usr/lib
/opt/freeware/lib
/opt/freeware/share/locale
/opt/freeware/lib
/opt/freeware/share/locale
```
Well, I just tried to replace /opt/freeware/lib with something of the same length:
`perl -pi -e 's#/opt/freeware/lib#/xxxxxxxxxxxxxxxx#g' libintl.a`
And that indeed made it try to import libiconv from /xxxxxxxxxxxxxxxx. So I think you are right!

Any idea how to edit libintl.a properly?

**Reply by rokm** [reply]
On linux, rpath can be modified using `chrpath` or `patchelf`. Perhaps either is available on your AIX via `dnf`?

**Reply by hurzhurz** [reply]
Unfortunately neither is available.

I have tried to compile patchelf, but just got it working half way ("make" worked, but "make check" did not finish).
So there is a binary "patchelf" but I'm not sure if something is missing.
Anyway, if I run it with --print-rpath against libintl.a or the extracted .so files, it always says:
`patchelf: not an ELF executable`

This is the output of `file`, if that matters:
```
libintl.a:    archive (big format)
libintl.so.1: executable (RISC System/6000 V3.1) or obj module not stripped
libintl.so.8: executable (RISC System/6000 V3.1) or obj module not stripped
```

**Reply by rokm** [reply]
No idea, then, sorry. AIX is well beyond my field of expertise.

**Reply by hurzhurz** [reply]
You already helped me a lot, thank you very much!
Now I know what to look for and maybe I will find a applicable way to change the rpath.
